crates/ra_assists/src/handlers/change_return_type_to_result.rs

```diff
@@ -0,0 +1,798 @@
+use ra_syntax::{
+    ast, AstNode,
+    SyntaxKind::{COMMENT, WHITESPACE},
+    SyntaxNode, TextSize,
+};
+
+use crate::{Assist, AssistCtx, AssistId};
+use ast::{Block, BlockExpr, Expr};
+
+// Assist: change_return_type_to_result
+//
+// Change the function's return type to Result.
+//
+// ```
+// <|>fn foo() -> i32 { 42i32 }
+// ```
+// ->
+// ```
+// fn foo() -> Result<i32, > { Ok(42i32) }
+// ```
+pub(crate) fn change_return_type_to_result(ctx: AssistCtx) -> Option<Assist> {
+    let fn_def = ctx.find_node_at_offset::<ast::FnDef>();
+    let fn_def = &mut fn_def?;
+    if fn_def.ret_type().is_none() {
+        return None;
+    }
+    let ret_type = &fn_def.ret_type().unwrap().type_ref()?;
+    if ret_type.syntax().text().to_string().starts_with("Result<") {
+        return None;
+    }
+
+    let mut tail_return_expr_collector = TailReturnCollector::new();
+    let fn_body = &fn_def.body()?;
+    let block_expr = &fn_body.block()?;
+
+    tail_return_expr_collector.collect_return_exprs(block_expr);
+    tail_return_expr_collector.collect_tail_exprs(block_expr);
```

Comment:
it doesn't seem like doing these things changes whether the assist is available, so we could move them into the `|edit|` closure?